vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc

```diff
@@ -828,45 +859,74 @@ void LedgerImpl::Shutdown(ledger::ResultCallback callback) {
   });
 }
 
-void LedgerImpl::OnAllDone(
-    const type::Result result,
-    ledger::ResultCallback callback) {
+void LedgerImpl::OnAllDone(type::Result result, ResultCallback callback) {
   database()->Close(callback);
 }
 
-void LedgerImpl::GetEventLogs(ledger::GetEventLogsCallback callback) {
-  database()->GetLastEventLogs(callback);
+void LedgerImpl::GetEventLogs(GetEventLogsCallback callback) {
+  WhenReady([this, callback]() { database()->GetLastEventLogs(callback); });
 }
 
 bool LedgerImpl::IsShuttingDown() const {
-  return shutting_down_;
+  return ready_state_ == ReadyState::kShuttingDown;
 }
 
 void LedgerImpl::GetBraveWallet(GetBraveWalletCallback callback) {
-  callback(wallet()->GetWallet());
+  WhenReady([this, callback]() { callback(wallet()->GetWallet()); });
 }
 
-std::string LedgerImpl::GetWalletPassphrase() const {
-  const auto brave_wallet = wallet()->GetWallet();
+std::string LedgerImpl::GetWalletPassphrase() {
+  if (!IsReady())
+    return "";
+
+  auto brave_wallet = wallet()->GetWallet();
   if (!brave_wallet) {
     return "";
   }
 
-  return wallet()->GetWalletPassphrase(brave_wallet->Clone());
+  return wallet()->GetWalletPassphrase(std::move(brave_wallet));
 }
 
 void LedgerImpl::LinkBraveWallet(const std::string& destination_payment_id,
                                  PostSuggestionsClaimCallback callback) {
-  wallet()->LinkBraveWallet(destination_payment_id, callback);
+  WhenReady([this, destination_payment_id, callback]() {
+    wallet()->LinkBraveWallet(destination_payment_id, callback);
+  });
 }
 
 void LedgerImpl::GetTransferableAmount(GetTransferableAmountCallback callback) {
-  promotion()->GetTransferableAmount(callback);
+  WhenReady(
+      [this, callback]() { promotion()->GetTransferableAmount(callback); });
 }
 
 void LedgerImpl::GetDrainStatus(const std::string& drain_id,
-                                ledger::GetDrainCallback callback) {
-  promotion()->GetDrainStatus(drain_id, callback);
+                                GetDrainCallback callback) {
+  WhenReady([this, drain_id, callback]() {
+    promotion()->GetDrainStatus(drain_id, callback);
+  });
+}
+
+bool LedgerImpl::IsReady() const {
+  return ready_state_ == ReadyState::kReady;
+}
+
+void LedgerImpl::WhenReady(std::function<void()> callback) {
+  switch (ready_state_) {
+    case ReadyState::kReady:
+      callback();
+      break;
+    default:
+      ready_callbacks_.push(callback);
+      break;
+  }
+}
+
+template <typename T>
```

Comment:
Since you use `WhenReady<>()` with rvalues only and since it can't be called with an lvalue, move-only lambda, e.g.:
```
void LedgerImpl::SaveRecurringTip(type::RecurringTipPtr info,
                                  ResultCallback callback) {
  auto lambda = [this, info = std::move(info), callback]() mutable {
    database()->SaveRecurringTip(std::move(info), callback);
  };
  WhenReady(lambda); // doesn't compile
}
```
you could just be explicit about it, use a forwarding reference and SFINAE away the lvalue reference overload:
```
template <typename T,
          typename = typename std::enable_if<!std::is_lvalue_reference<T>::value>::type>
void WhenReady(T&& callback) {
  WhenReady_impl(
      [shared_callback = std::make_shared<T>(std::move(callback))]() {
        (*shared_callback)();
      });
}
```
(and rename the non-template overload).

Reply:
Thanks! I think in this case I would prefer to not use SFINAE, since devs are pretty used to seeing and dealing with "copy constructor deleted" compile errors in cases like this. I did make a change to skip the unnecessary `std::function` wrapper when the callback can be executed directly, though.